Add DETA

NielsRogge opened the pull request:
# What does this PR do?

This PR adds [DETA](https://github.com/jozhang97/DETA/issues/3). DETA is a slight change to Deformable DETR by using traditional IoU-based assignment as opposed to the Hungarian matching used in the original DETR, and incorporating NMS (non-maximum suppression) in the postprocessing.

Note: this model has a `torchvision` dependency for NMS.

To do:

- [x] transfer checkpoints

NielsRogge committed:
First draft

NielsRogge committed:
Add initial draft of conversion script

NielsRogge committed:
Convert all weights

NielsRogge committed:
Fix config

NielsRogge committed:
Add image processor

NielsRogge committed:
Fix DetaImageProcessor

NielsRogge committed:
Run make fix copies

NielsRogge committed:
Remove timm dependency

NielsRogge committed:
Fix dummy objects

NielsRogge committed:
Improve loss function

NielsRogge committed:
Remove conv_encoder attribute

NielsRogge committed:
Update conversion scripts

NielsRogge committed:
Improve postprocessing + docs

NielsRogge committed:
Fix copied from statements

NielsRogge committed:
Add tests

NielsRogge committed:
Improve postprocessing

NielsRogge committed:
Improve postprocessing

NielsRogge committed:
Update READMEs

NielsRogge committed:
More improvements

NielsRogge committed:
Fix rebase

NielsRogge commented:
cc @alaradirik this PR is in a ready state, except for 2 things:

- [ ] whether or not we leverage torchvision's `batched_nms` => the CI is currently failing because this library is not installed. Will also ask for @sgugger and @LysandreJik's opinion here
- [ ] the `post_process_object_detection` method might require an in-depth look

sgugger commented:
There is no problem with the model requiring torchvision to be installed. We have many models with specific dependencies, some of which you ported yourself ;-).
Just protect the import between `if is_torchvision_available()` and have a the first line in the init of the models be a `require_backends(["torchvision"])`.

NielsRogge committed:
Add is_torchvision_available

NielsRogge committed:
Add torchvision dependency

NielsRogge committed:
Fix typo and README

alaradirik reviewed (commented):
Thanks for adding this! I added a few comments but it looks good to me overall.

Could you add an `nms_threshold` argument to the object detection post processing? I can do a follow up PR to add it to all other post processing methods. Or it can be left as it is and I can add it shortly and update the object detection pipeline as well

NielsRogge commented on src/transformers/models/deta/image_processing_deta.py:
```diff
@@ -0,0 +1,1323 @@
+        Args:
+            outputs ([`DetrObjectDetectionOutput`]):
+                Raw outputs of the model.
+            threshold (`float`, *optional*):
```
We can't just add NMS support to all post_process_object_detection methods I'm afraid, since for that one uses Torchvision's NMS op which is written in C and much faster than plain python. Also our existing models don't need NMS, DETA is actually the first one that needs it

alaradirik replied:
NMS is a quite common to post-processing method for object detection and useful for models with noisy bounding box proposals. Given two bounding boxes with high prediction scores, they might be detecting the same object instance and NMS eliminates the redundant boxes.

I think it'd be nice to have an option to perform NMS but I'm fine with not adding it as well. 
CC'ing @amyeroberts and @sgugger

amyeroberts replied:
> since for that one uses Torchvision's NMS op which is written in C and much faster than plain python

@NielsRogge I'm not sure I completely follow - is the issue here speed or the use of torchvision? 

In general, I agree with @alaradirik, NMS is common enough that it's something I think we want to support, even if it's only for our PyTorch models at the moment.

NielsRogge replied:
Ok we can add NMS as an option, this way the API of all `post_process_object_detection` methods will be the same. I assume we'll want to leverage torchvision for this

alaradirik replied:
Yes, it makes sense to leverage torchvision for this

NielsRogge committed:
Fix bug

NielsRogge committed:
Add copied from

NielsRogge committed:
Fix style

NielsRogge commented on src/transformers/models/deta/modeling_deta.py:
```diff
@@ -0,0 +1,2815 @@
+        Classification loss (Binary focal loss) targets dicts must contain the key "class_labels" containing a tensor
+        of dim [nb_target_boxes]
+        """
+        if "logits" not in outputs:
+            raise KeyError("No logits were found in the outputs")
```
This is copied from DETR. Why would you remove this?

NielsRogge replied:
As it's copied, I'll keep it (we can improve this in a separate PR)

NielsRogge commented on src/transformers/models/deta/modeling_deta.py:
```diff
@@ -0,0 +1,2815 @@
+                    is_level_ordered = (
+                        level_ids[keep_inds][None]
+                        == torch.arange(len(spatial_shapes), device=level_ids.device)[:, None]
+                    )  # LS
```
Pinging the authors here, @xingyizhou @jozhang97, could you clarify what LS means here?

NielsRogge committed:
Apply suggestions

NielsRogge committed:
Fix thanks to @ydshieh

NielsRogge committed:
Fix another dependency check

NielsRogge committed:
Simplify image processor

NielsRogge committed:
Add scipy

NielsRogge committed:
Improve code

NielsRogge commented on src/transformers/models/deta/modeling_deta.py:
```diff
@@ -0,0 +1,2815 @@
+                    prop_logits_b = proposal_logit[b]
+
+                    # pre-nms per-level topk
+                    pre_nms_topk = 1000
```
Yes this is hardcoded here: https://github.com/jozhang97/DETA/blob/dade1763efba58a1f3077d373e991fd319dc240e/models/deformable_transformer.py#L184

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_20983). All of your documentation changes will be reflected on that endpoint.

NielsRogge committed:
Add threshold argument

NielsRogge committed:
Fix bug

NielsRogge committed:
Set default threshold

NielsRogge commented on src/transformers/models/deta/modeling_deta.py:
```diff
@@ -0,0 +1,2815 @@
+
+@add_start_docstrings(
+    """
+    DETA Model (consisting of a backbone and encoder-decoder Transformer) with object detection heads on top, for tasks
+    such as COCO detection.
```
We usually refer to benchmarks, so I'll keep COCO if you don't mind :)

NielsRogge committed:
Improve integration test

sgugger reviewed (commented):
Thanks for adding this new model. Left a couple of comments.

NielsRogge committed:
Add another integration test

NielsRogge committed:
Update setup.py

NielsRogge committed:
Address review

NielsRogge commented on src/transformers/models/deta/image_processing_deta.py:
```diff
@@ -0,0 +1,1041 @@
+        do_pad: bool = True,
+        **kwargs
+    ) -> None:
+        if "pad_and_return_pixel_mask" in kwargs:
+            do_pad = kwargs.pop("pad_and_return_pixel_mask")
+
+        if "max_size" in kwargs:
+            warnings.warn(
+                "The `max_size` parameter is deprecated and will be removed in v4.26. "
+                "Please specify in `size['longest_edge'] instead`.",
+                FutureWarning,
+            )
+            max_size = kwargs.pop("max_size")
+        else:
+            max_size = None if size is None else 1333
```
Ok, however this is due to use of copied from statements.

Should I remove the copied from to remove the deprecation warnings?

sgugger replied:
Yes you can remove the copied from statement for the init to be able to remove those.

NielsRogge committed:
Improve deformable attention function

NielsRogge committed:
Improve copied from

NielsRogge committed:
Use relative imports

NielsRogge committed:
Address review

NielsRogge committed:
Replace assertions

NielsRogge committed:
Address review

NielsRogge committed:
Update dummies

NielsRogge commented on src/transformers/__init__.py:
```diff
@@ -855,6 +859,25 @@
+            "DetaModel",
+            "DetaPreTrainedModel",
+        ]
+    )
```
Ok, switched to `is_torch_available()`

NielsRogge committed:
Remove dummies

NielsRogge committed:
Address comments, update READMEs

NielsRogge commented:
@sgugger I've addressed all comments, except for adding support for the custom kernel.

Could we perhaps add support for the custom kernel for the 3 models (Mask2Former, OneFormer and DETA) in a separate PR?

sgugger commented:
In this case, remove the code trying to load the custom kernels in the modeling file and we can add it back in the PR that will deal with custom kernels.

NielsRogge committed:
Remove custom kernel code

NielsRogge committed:
Add image processor tests

NielsRogge commented:
@sgugger ok, feel free to approve :)

sgugger commented on src/transformers/models/deta/modeling_deta.py:
```diff
@@ -0,0 +1,2755 @@
+        super().__init__(config)
+
+        # Create backbone with positional encoding
+        self.backbone = DetaBackboneWithPositionalEncodings(config)
```
This is missing a `requires_backends(self, ["torchvision"])` since this is where the `bached_nms` is used.

sgugger reviewed (approved):
Thanks for all your work on this!

NielsRogge committed:
Add requires_backends

NielsRogge committed:
Add minor comment

NielsRogge committed:
Update scripts

NielsRogge committed:
Update organization name

NielsRogge committed:
Fix defaults, add doc tests

NielsRogge committed:
Add id2label for object 365

NielsRogge committed:
Fix tests

NielsRogge committed:
Update task guide

NielsRogge commented:
Failing test is unrelated/flaky, merging.
